File: fud-ai/android/app/src/main/java/com/apoorvdarshan/calorietracker/services/WidgetSnapshotWriter.kt
package com.apoorvdarshan.calorietracker.services

import android.content.Context
import androidx.compose.ui.graphics.toArgb
import com.apoorvdarshan.calorietracker.data.FoodRepository
import com.apoorvdarshan.calorietracker.data.PreferencesStore
import com.apoorvdarshan.calorietracker.data.ProfileRepository
import com.apoorvdarshan.calorietracker.models.FoodEntry
import com.apoorvdarshan.calorietracker.models.HomeTopNutrient
import com.apoorvdarshan.calorietracker.models.UserProfile
import com.apoorvdarshan.calorietracker.models.WidgetNutrient
import com.apoorvdarshan.calorietracker.models.WidgetSnapshot
import com.apoorvdarshan.calorietracker.models.WaterEntry
import com.apoorvdarshan.calorietracker.models.WaterUnit
import com.apoorvdarshan.calorietracker.ui.theme.AppThemeColor
import com.apoorvdarshan.calorietracker.widget.WidgetRefreshScheduler
import com.apoorvdarshan.calorietracker.widget.WidgetUpdateCoordinator
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onEach
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

/**
 * Recomputes today's totals whenever food entries or the user profile change,
 * writes the [WidgetSnapshot] into DataStore, and asks Glance to redraw both
 * the Calorie and Protein app widgets. Mirrors the iOS WidgetSnapshotWriter
 * call sites (every FoodStore change + profile-change notification + scene
 * resume).
 */
class WidgetSnapshotWriter(
    private val context: Context,
    private val prefs: PreferencesStore,
    private val foodRepository: FoodRepository,
    private val profileRepository: ProfileRepository
) {
    fun observe() = combine(
        combine(
            foodRepository.entries,
            profileRepository.profile,
            prefs.homeTopNutrients,
            prefs.appThemeColor,
            prefs.optionalNutrientGoals
        ) { entries, profile, _, _, _ ->
            // Selection / theme / goals are re-read inside publish; they're combined
            // here only so their changes re-trigger a snapshot write.
            CoreWidgetState(entries, profile)
        },
        combine(
            prefs.waterTrackingEnabled,
            prefs.waterDailyGoalMl,
            prefs.waterUnit,
            prefs.waterEntries
        ) { enabled, goalMl, unit, entries ->
            WaterWidgetState(enabled, goalMl, unit, entries)
        }
    ) { core, water -> WidgetInputs(core, water) }
        .distinctUntilChanged()
        .onEach { inputs -> publish(inputs.core.entries, inputs.core.profile, inputs.water) }
        .map { Unit }

    private suspend fun publish(entries: List<FoodEntry>, profile: UserProfile?, water: WaterWidgetState) {
        val todaysEntries = entries.filter {
            it.timestamp.atZone(ZoneId.systemDefault()).toLocalDate() == LocalDate.now()
        }
        if (profile == null) {
            prefs.clearWidgetSnapshot()
        } else {
            val selection = HomeTopNutrient.fromStorage(prefs.homeTopNutrients.first())
            val optionalGoals = prefs.optionalNutrientGoals.first()
            val theme = AppThemeColor.fromKey(prefs.appThemeColor.first())
            val waterTodayMl = water.entries
                .filter { it.date.atZone(ZoneId.systemDefault()).toLocalDate() == LocalDate.now() }
                .sumOf { it.milliliters }
            val snapshot = WidgetSnapshot(
                date = Instant.now(),
                dayStart = WidgetSnapshot.todayStart(),
                calories = todaysEntries.sumOf { it.calories },
                calorieGoal = profile.effectiveCalories,
                protein = todaysEntries.sumOf { it.protein },
                proteinGoal = profile.effectiveProtein,
                carbs = todaysEntries.sumOf { it.carbs },
                carbsGoal = profile.effectiveCarbs,
                fat = todaysEntries.sumOf { it.fat },
                fatGoal = profile.effectiveFat,
                homeNutrients = selection.map { nutrient ->
                    WidgetNutrient(
                        id = nutrient.storageKey,
                        label = context.getString(nutrient.displayNameRes),
                        unit = context.getString(nutrient.unitRes),
                        value = nutrient.current(todaysEntries),
                        goal = nutrient.goal(profile, optionalGoals).toDouble()
                    )
                },
                themeStartHex = theme.start.toArgb() and 0xFFFFFF,
                themeEndHex = theme.end.toArgb() and 0xFFFFFF,
                waterTrackingEnabled = water.enabled,
                waterCurrentMl = waterTodayMl,
                waterGoalMl = water.goalMl.coerceAtLeast(1),
                waterUnitRaw = water.unit.storageValue
            )
            prefs.setWidgetSnapshot(snapshot)
        }
        if (!WidgetUpdateCoordinator.updateAll(context)) {
            WidgetRefreshScheduler.enqueueImmediate(context, "snapshot_update_failed")
        }
    }
}

private data class CoreWidgetState(
    val entries: List<FoodEntry>,
    val profile: UserProfile?
)

private data class WaterWidgetState(
    val enabled: Boolean,
    val goalMl: Int,
    val unit: WaterUnit,
    val entries: List<WaterEntry>
)

private data class WidgetInputs(
    val core: CoreWidgetState,
    val water: WaterWidgetState
)

File: fud-ai/ios/calorietracker/Services/WidgetSnapshotWriter.swift
import Foundation
import WidgetKit

/// Writes a WidgetSnapshot into the shared App Group container and asks
/// WidgetKit to refresh all timelines. Widgets can't read the main app's
/// private UserDefaults, so any data the widget needs has to go through here.
enum WidgetSnapshotWriter {
    /// Recomputes today's totals from the current FoodStore + ProfileStore and
    /// publishes them to the widget.
    static func publish(foods: [FoodEntry], profile: UserProfile) {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: Date())
        // Must be same-day — `timestamp >= startOfDay` alone would fold future-logged
        // entries (meals planned on tomorrow via the week strip) into today's totals.
        let today = foods.filter { calendar.isDate($0.timestamp, inSameDayAs: Date()) }

        let cal = today.reduce(0) { $0 + $1.calories }
        let p = today.reduce(0) { $0 + $1.protein }
        let c = today.reduce(0) { $0 + $1.carbs }
        let f = today.reduce(0) { $0 + $1.fat }
        let selectedHomeNutrients = HomeTopNutrient.selection(
            from: UserDefaults.standard.string(forKey: HomeTopNutrient.storageKey)
                ?? HomeTopNutrient.storageValue(for: HomeTopNutrient.defaultSelection)
        )
        let optionalGoals = OptionalNutrientGoals.current
        let theme = AppThemeColor(
            rawValue: UserDefaults.standard.string(forKey: AppThemeColor.storageKey) ?? ""
        ) ?? .defaultColor
        let waterEntries: [WaterEntry] = UserDefaults.standard.data(forKey: WaterSettings.entriesKey)
            .flatMap { try? JSONDecoder().decode([WaterEntry].self, from: $0) } ?? []
        let waterCurrentMl = waterEntries
            .filter { calendar.isDate($0.date, inSameDayAs: Date()) }
            .reduce(0) { $0 + $1.milliliters }
        let storedWaterGoal = UserDefaults.standard.integer(forKey: WaterSettings.dailyGoalKey)
        let waterGoalMl = storedWaterGoal > 0 ? storedWaterGoal : WaterSettings.defaultDailyGoalMl

        let snapshot = WidgetSnapshot(
            date: Date(),
            dayStart: startOfDay,
            calories: cal,
            calorieGoal: profile.effectiveCalories,
            protein: p,
            proteinGoal: profile.effectiveProtein,
            carbs: c,
            carbsGoal: profile.effectiveCarbs,
            fat: f,
            fatGoal: profile.effectiveFat,
            homeNutrients: selectedHomeNutrients.map {
                homeNutrientValue(for: $0, foods: today, profile: profile, optionalGoals: optionalGoals)
            },
            waterTrackingEnabled: UserDefaults.standard.bool(forKey: WaterSettings.enabledKey),
            waterCurrentMl: waterCurrentMl,
            waterGoalMl: waterGoalMl,
            waterUnitRaw: UserDefaults.standard.string(forKey: WaterSettings.unitKey) ?? WaterUnit.defaultUnit.rawValue,
            themeStartHex: theme.startHex,
            themeEndHex: theme.endHex
        )

        if WidgetSnapshot.read() != snapshot {
            WidgetSnapshot.write(snapshot)
            WidgetCenter.shared.reloadAllTimelines()
        }
        WatchSnapshotSync.shared.send(snapshot)
    }

    private static func homeNutrientValue(
        for nutrient: HomeTopNutrient,
        foods: [FoodEntry],
        profile: UserProfile,
        optionalGoals: OptionalNutrientGoals
    ) -> WidgetNutrientValue {
        WidgetNutrientValue(
            id: nutrient.rawValue,
            label: nutrient.displayName,
            shortLabel: shortLabel(for: nutrient),
            unit: nutrient.unit,
            iconName: nutrient.iconName,
            value: currentValue(for: nutrient, foods: foods),
            goal: nutrient.goal(for: profile, optionalGoals: optionalGoals)
        )
    }

    private static func currentValue(for nutrient: HomeTopNutrient, foods: [FoodEntry]) -> Double {
        switch nutrient {
        case .protein:
            return Double(foods.reduce(0) { $0 + $1.protein })
        case .carbs:
            return Double(foods.reduce(0) { $0 + $1.carbs })
        case .fat:
            return Double(foods.reduce(0) { $0 + $1.fat })
        case .fiber:
            return sum(foods, \.fiber)
        case .sugar:
            return sum(foods, \.sugar)
        case .addedSugar:
            return sum(foods, \.addedSugar)
        case .saturatedFat:
            return sum(foods, \.saturatedFat)
        case .cholesterol:
            return sum(foods, \.cholesterol)
        case .sodium:
            return sum(foods, \.sodium)
        case .potassium:
            return sum(foods, \.potassium)
        case .transFat:
            return sum(foods, \.transFat)
        case .calcium:
            return sum(foods, \.calcium)
        case .iron:
            return sum(foods, \.iron)
        case .magnesium:
            return sum(foods, \.magnesium)
        case .zinc:
            return sum(foods, \.zinc)
        case .vitaminA:
            return sum(foods, \.vitaminA)
        case .vitaminC:
            return sum(foods, \.vitaminC)
        case .vitaminD:
            return sum(foods, \.vitaminD)
        case .vitaminB12:
            return sum(foods, \.vitaminB12)
        case .vitaminE:
            return sum(foods, \.vitaminE)
        case .vitaminK:
            return sum(foods, \.vitaminK)
        case .folate:
            return sum(foods, \.folate)
        case .omega3:
            return sum(foods, \.omega3)
        }
    }

    private static func sum(_ foods: [FoodEntry], _ keyPath: KeyPath<FoodEntry, Double?>) -> Double {
        foods.reduce(0) { $0 + ($1[keyPath: keyPath] ?? 0) }
    }

    private static func shortLabel(for nutrient: HomeTopNutrient) -> String {
        String(nutrient.displayName.prefix(1)).uppercased()
    }
}
